### PR TITLE
qualifiers in meta knowledge graph

### PR DESCRIPTION
**Pull request by rjawesome**
https://github.com/biothings/BioThings_Explorer_TRAPI/issues/595
Qualifiers are just read from the existing qualifiers in yaml.

Currently doesn't merge edges with same input/output/predicate but different qualfiers in the meta knowledge graph.

### CODE REVIEW

File: src/controllers/meta_knowledge_graph.js [hunk 1]
```diff
@@ -80,7 +82,7 @@ module.exports = class MetaKnowledgeGraphHandler {
         predicates[input][output] = [];
       }
       if (!predicates[input][output].includes(pred)) {
```
Comment (tokebe) [inline src/controllers/meta_knowledge_graph.js]: This check should be updated to ensure no duplicates are added

**Review by tokebe — CHANGES_REQUESTED** [review]
So far so good, but a few notes:
- `qualifier_type_id` should have the `biolink:` prefix stripped and re-added to ensure it's present and non-duplicated
- applicable_values should have `biolink:` prefix stripped, and only added if the `qualifier_type_id` includes `"predicate"`

Regarding your question about `supportedLookups()`:

`supportedLookups()` should add the complete qualifier set from the given template (provided the template has qualifiers) to each supported lookup it generates. In `controllers/meta_knowledge_graph.js`, the step that calls `supportedLookups()` should use this new qualifier information alongside the subject/predicate/object to attempt to match to an existing edge. If it matches, add the knowledge type inferred to that edge, otherwise make a new edge with just knowledge type inferred and the information from the supported lookup.

**Comment by rjawesome** [comment]
@tokebe Can you give a sample template file. I don't see it in the bte trapi repo but maybe I'm missing it?

**Comment by tokebe** [comment]
Follow the imports: `controllers/meta_knowledge_graphs.js` imports `supportedLookups()` from [`query_graph_handler`](https://github.com/biothings/bte_trapi_query_graph_handler/blob/380ee0691bd00a92605450aedb9528a13ed1fe11/src/index.js#L27), which imports it from [`query_graph_handler/src/inferred_mode/template_lookup.js`](https://github.com/biothings/bte_trapi_query_graph_handler/blob/380ee0691bd00a92605450aedb9528a13ed1fe11/src/inferred_mode/template_lookup.js#L54), which shows that templates are read from [`templateGroups.json`](https://github.com/biothings/bte_trapi_query_graph_handler/blob/main/data/templateGroups.json).

You'll need to edit the function in template_lookup.js as well to actually grab the qualifier sets.

**Comment by rjawesome** [comment]
Oh I was looking for the json file in the wrong repo...

**Comment by rjawesome** [comment]
Should be fixed, also a new PR on query graph handler

**Comment by colleenXu** [comment]
Regarding `association` functionality...
* Could we put it into a separate PR / separate set of PRs? I'm thinking that we'll test this functionality later, when we put the info into the SmartAPI yamls
* could the SmartAPI yaml (x-bte operation) property name be changed from "association_id" -> "biolink_association_type"?
* In the SmartAPI yaml, the property's value will probably be just the term like [EntityToDiseaseAssociation](https://biolink.github.io/biolink-model/docs/EntityToDiseaseAssociation.html). But in the endpoint response, we will want this to have the biolink-prefix (`"association": "biolink:EntityToDiseaseAssociation"`)

**Comment by colleenXu** [comment]
@tokebe 

I have a question specific to the `v1/meta_knowledge_graph` response. When a MetaEdge can be done as `lookup` or creative-mode....did we want to have one operation with `"knowledge_types": ["inferred", "lookup"]` or two? 

With the current PR code, two separate MetaEdges are made. I saw these two:

```
        {
            "subject": "biolink:SmallMolecule",
            "predicate": "biolink:treats",
            "object": "biolink:Disease",
            "knowledge_types": [
                "lookup"
            ]
        },


        {
            "subject": "biolink:SmallMolecule",
            "predicate": "biolink:treats",
            "object": "biolink:Disease",
            "knowledge_types": [
                "inferred"
            ]
        },
```

**Comment by colleenXu** [comment]
@tokebe @rjawesome 

Regarding `association` functionality...
* Could we put it into a separate PR / separate set of PRs? I'm thinking that we'll test this functionality later, when we put the info into the SmartAPI yamls
* could the SmartAPI yaml (x-bte operation) property name be changed from "association_id" -> "biolink_association_type"?
* Note: In the SmartAPI yaml, the property's value will probably be just the term like [EntityToDiseaseAssociation](https://biolink.github.io/biolink-model/docs/EntityToDiseaseAssociation.html). But in the endpoint response, we will want this to have the biolink-prefix (`"association": "biolink:EntityToDiseaseAssociation"`)

**Comment by colleenXu** [comment]
And I changed the title because I assumed we meant "qualifiers". 

I'm done with my quick testing of the "qualifiers functionality" for https://github.com/biothings/BioThings_Explorer_TRAPI/issues/595. I posed my two questions / chunks of feedback above, and I guess we'll wait to hear about the "qualifier-set merging" from the TRAPI team / Translator (and we'll track that discussion in the issue). 

It looks great overall!

**Comment by rjawesome** [comment]
Association has been moved into its own PR. Will address Colleen's comments in that PR

**Comment by tokebe** [comment]
> @tokebe
> 
> I have a question specific to the `v1/meta_knowledge_graph` response. When a MetaEdge can be done as `lookup` or creative-mode....did we want to have one operation with `"knowledge_types": ["inferred", "lookup"]` or two?
> 
> With the current PR code, two separate MetaEdges are made. I saw these two:
> 
> ```
>         {
>             "subject": "biolink:SmallMolecule",
>             "predicate": "biolink:treats",
>             "object": "biolink:Disease",
>             "knowledge_types": [
>                 "lookup"
>             ]
>         },
> 
> 
>         {
>             "subject": "biolink:SmallMolecule",
>             "predicate": "biolink:treats",
>             "object": "biolink:Disease",
>             "knowledge_types": [
>                 "inferred"
>             ]
>         },
> ```

We want edges that are exactly the same other than `knowledge_types` to be merged, so it's one edge with `"knowledge_types": ["inferred", "lookup"]`.

**Comment by rjawesome** [comment]
Should be fixed

**Comment by tokebe** [comment]
Fix currently has inferred duplicating on some edges, like below:

```json
{
      "subject": "biolink:SmallMolecule",
      "predicate": "biolink:affects",
      "object": "biolink:Gene",
      "qualifiers": [
        {
          "qualifier_type_id": "biolink:causal_mechanism_qualifier",
          "applicable_values": [
            "inverse_agonism"
          ]
        },
        {
          "qualifier_type_id": "biolink:object_aspect_qualifier",
          "applicable_values": [
            "activity"
          ]
        },
        {
          "qualifier_type_id": "biolink:object_direction_qualifier",
          "applicable_values": [
            "decreased"
          ]
        },
        {
          "qualifier_type_id": "biolink:qualified_predicate",
          "applicable_values": [
            "biolink:causes"
          ]
        }
      ],
      "knowledge_types": [
        "lookup",
        "inferred",
        "inferred"
      ]
    },
```

**Comment by tokebe** [comment]
(Also, I merged the latest from main, you'll have to update your workspace by cloning this repo fresh and using `npm run setup`)

**Comment by rjawesome** [comment]
Shoudl be fixed

**Comment by colleenXu** [comment]
Hmmm....What is shown [above](https://github.com/biothings/biothings_explorer/pull/610#issuecomment-1508735715) is actually "incorrect" because we wouldn't actually run creative-mode on a QEdge if all those qualifiers were specified as constraints on that QEdge. 

I think some logic may be needed to figure out when a `lookup` TRAPI MetaEdge should have the `inferred` string added to `knowledge_types` (aka "merging"). Maybe the logic is similar to what's done to match a QGraph QEdge to creative-mode templates...

Some thoughts I have: 
* if a `lookup` edge has more qualifiers specified (larger set of type_ids), then it is representing MORE SPECIFIC stuff than the creative-mode template and creative-mode won't run. So we shouldn't add the `inferred` string...
* If the `lookup` edge had the exact same set of qualifier type_ids/values as a creative-mode template (and the subject/predicate/object matches too), then we want to add the `inferred` string
    * I doubt that we have a `lookup` edge for `SmallMolecule-affects-Gene` with the exact qualifier set `increased activity_or_abundance`. But if we do, then we'd want to add the `inferred` string. 
    * If we don't have that `lookup` edge, we'll make an `inferred`-only TRAPI MetaEdge...
* but we also currently have some ambiguity in our creative-mode behavior: creative-mode will happen if the `inferred` QEdge's qualifier set is more general. Do we want to add the `inferred` string in these situations?
    *  For example, I know we have a `lookup` edge where `SmallMolecule-affects-Gene` with NO qualifiers. We could decide to add the `inferred` string there, since creative-mode will run if an `inferred` QEdge like that was provided...
    * Another example: I doubt that we have a `lookup` edge SmallMolecule-affects-Gene` with the exact qualifier set `increased`. But if we did, we could decide to add the `inferred` string there since creative mode would run...
    * I'm not sure we ever got clarification from the rest of Translator on whether this ambiguity in behavior was okay. We discussed it as "point 2" earlier this year here: my [question](https://github.com/biothings/biothings_explorer/issues/532#issuecomment-1378322378), Jackson [reply](https://github.com/biothings/biothings_explorer/issues/532#issuecomment-1378953156), Andrew [reply](https://github.com/biothings/biothings_explorer/issues/532#issuecomment-1379094937)

**Comment by tokebe** [comment]
@colleenXu I'm not convinced this logic is necessary... based on the way the poll is going, it's looking like Translator want the more general "all possible qualifiers" option, in which case such ambiguity is expected. It wouldn't be read as "we support creative mode on this edge with these qualifiers", so much as "we support creative mode on this edge with some subset of these qualifiers" and then we have no duty to be more clear.

**Comment by tokebe** [comment]
Looks like this is almost ready to go -- I'm noticing a couple cases of duplicate "inferred" in some of the metaEdges, however:

```
{
      "subject": "biolink:ComplexMolecularMixture",
      "predicate": "biolink:affects",
      "object": "biolink:Protein",
      "qualifiers": [
        {
          "qualifier_type_id": "biolink:object_aspect_qualifier",
          "applicable_values": ["activity_or_abundance"]
        },
        {
          "qualifier_type_id": "biolink:object_direction_qualifier",
          "applicable_values": ["increased", "decreased"]
        }
      ],
      "knowledge_types": ["inferred", "inferred"]
    }
```

**Review by tokebe — APPROVED** [review]

**Pull request merged**